File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
def dfs(v, visited, graph):
    visited[v] = True
    for i in graph[v]:
        if not visited[i]:
            dfs(i, visited, graph)

input_values = input().split()
n = int(input_values[0])
m = int(input_values[1])

graph = [[] for _ in range(n + 1)]
visited = [False for _ in range(n + 1)]

for _ in range(m):
    s, e = map(int, input().split())
    graph[s].append(e)
    graph[e].append(s)

count = 0
for i in range(1, n + 1):
    if not visited[i]:
        count += 1
        dfs(i, visited, graph)

print(count)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.rs
use std::io::{BufReader,BufWriter,BufRead,Write,stdin,stdout};
fn dfs(v: usize, visited: &mut Vec<bool>, graph: &Vec<Vec<usize>>) {
    visited[v] = true;
    for &i in &graph[v] {
        if !visited[i] {
            dfs(i, visited, graph);
        }
    }
}

fn main() {
    let mut reader= BufReader::new(stdin().lock());
    let mut writer = BufWriter::new(stdout().lock());
    let mut input = String::new();    
    reader.read_line(&mut input).unwrap();
    let  inputs:Vec<usize>= input.trim().split_ascii_whitespace().map(|x|x.parse::<usize>().unwrap()).collect();
    let n: usize = inputs[0];
    let m: usize =inputs[1];

    let mut graph: Vec<Vec<usize>> = vec![vec![]; n  + 1];
    let mut visited: Vec<bool> = vec![false; n  + 1];

    for _ in 0..m {
        input.clear();
        reader.read_line(&mut input).unwrap();
        let  inputs:Vec<usize>= input.trim().split_ascii_whitespace().map(|x|x.parse::<usize>().unwrap()).collect();
        let s: usize = inputs[0];
        let e: usize =inputs[1];
    
        graph[s ].push(e);
        graph[e ].push(s);
    }

    let mut count = 0;
    for i in 1..=n {
        if !visited[i ] {
            count += 1;
            dfs(i , &mut visited, &graph);
        }
    }

    writeln!(writer,"{}", count).unwrap();
    writer.flush().unwrap();
}
